mesonpy/__init__.py

```diff
@@ -871,45 +871,13 @@ def sdist(self, directory: Path) -> pathlib.Path:
 
         with tarfile.open(meson_dist_path, 'r:gz') as meson_dist, mesonpy._util.create_targz(sdist) as tar:
             for member in meson_dist.getmembers():
-                # calculate the file path in the source directory
-                assert member.name, member.name
-                member_parts = member.name.split('/')
-                if len(member_parts) <= 1:
-                    continue
-                path = self._source_dir.joinpath(*member_parts[1:])
-
-                if not path.exists() and member.isfile():
-                    # File doesn't exists on the source directory but exists on
-                    # the Meson dist, so it is generated file, which we need to
-                    # include.
-                    # See https://mesonbuild.com/Reference-manual_builtin_meson.html#mesonadd_dist_script
-
-                    # MESON_DIST_ROOT could have a different base name
-                    # than the actual sdist basename, so we need to rename here
+                if member.isfile():
                     file = meson_dist.extractfile(member.name)
-                    member.name = str(pathlib.Path(dist_name, *member_parts[1:]).as_posix())
+                    # rewrite the path to match the sdist distribution name
+                    member.name = '/'.join((dist_name, *member.name.split('/')[1:]))
                     tar.addfile(member, file)
                     continue
 
-                if not path.is_file():
-                    continue
-
-                info = tarfile.TarInfo(member.name)
-                file_stat = os.stat(path)
-                info.mtime = member.mtime
-                info.size = file_stat.st_size
-                info.mode = int(oct(file_stat.st_mode)[-3:], 8)
```

Comment:
These changes are meaningful and not strictly related to uncommitted changes to files.

Example with SciPy:
```
$ ls -l tools/lint.*  # local git repo
-rwxr-xr-x 1 rgommers rgommers 4318 19 apr 12:16 ../tools/lint.py
-rw-r--r-- 1 rgommers rgommers 1180 19 apr 12:16 ../tools/lint.toml
```

Compared to what I see in the sdist (screenshots because extracting files changes ownership):

with 0.16.0:
<img width="498" alt="image" src="https://github.com/mesonbuild/meson-python/assets/98330/d54a7538-56cc-4b71-88f7-04bd85a5d08b">


with this PR:
<img width="498" alt="image" src="https://github.com/mesonbuild/meson-python/assets/98330/d22b9f39-2940-4ac7-ac69-9ae00c167590">

Note that the file permissions changed compared to what's in the git repo, and ownership now includes my own username. The former I'm not sure about either way, that is behavior inherited from `meson dist` it looks like. The latter seems undesirable.

Reply:
The change is unintended, in the sense that I trusted`meson dist` (and thus in turn `git archive`) to be doing the right thing. I'm also not sure whether the previous behavior was deliberate: it seems that files generated via dist scripts do not go through the same metadata mangling, and the metadata mangling seems more an accidental omission than a normalization.

Reply:
For file ownership, I checked `git archive --format=tar.gz HEAD > tmp.tar.gz` and it sets the ownership to `root`. So I guess the repacking changes it to local-username. Either way, the current behavior seems preferred, since it gives the same result when run on multiple machines.

Re file permissions: there are config settings for it, e.g. `tar.umask` in https://git-scm.com/docs/git-archive#_configuration. What `git-archive` does by default is probably good for reproducibility, and seems to be by design. 

Having a peek in `meson/mdist.py`, it doesn't pass any options like `--owner=0`/`--group=0`, `--no-same-owner`, `--no-same-permissions` to `tar`. Not sure if that just never came up, or was rejected before. I can't find anything related in the Meson issue tracker so quickly.

> I'm also not sure whether the previous behavior was deliberate: it seems that files generated via dist scripts do not go through the same metadata mangling

I wouldn't read too much into that, since this code was written pretty early in the project's history, and I don't think I knew about `add_dist_script` at the time; use of `add_dist_script` probably hadn't come up at all yet.

Reply:
I only spent a few minutes investigating the thing, and it seems that the archive metadata changes happen in Meson. I tend to think that it is a Meson bug. Unless there are subtleties I am missing, meson-python just copies over the archive members as is. The Meson behavior needs to be fixed, but in the meantime we can fix the metadata in meson-python.

> What `git-archive` does by default is probably good for reproducibility, and seems to be by design.

What does `git archive` do, other than setting the owner to root:root?

Reply:
One funny thing, is that `git archive`, Meson, and meson-python seem to use three different tar formats.

It would be nice if there were a way to avoid packing the files up three times, but I'm not sure there is one.

Reply:
Meson uses shutil.make_archive, which isn't well known for it's high degree of configurability. No `tar` dependency, dist works on Windows too.

The permissions are likely caused by the user umask setting.

Note that reproducibility is not necessarily a requirement of `meson dist` -- while avoiding pointless differences could be a "nice to have", there is no getting around the fact that a dist script can do *literally* whatever it wants and meson's job is to faithfully archive the result. This may mean file contents of a file in the resulting dist, have different text (the simplest example being dates, which makes sense since e.g. manpages should preferably have the date of the release, which isn't necessarily the same as the last commit and is most probably not the date the build was performed).

Since releases are typically done once and uploaded to static file hosting, to be downloaded many times, there is little value in reproducibility beyond auditability of git -> dist. And the dangers there are things like extremely noisy file content changes, not really file metadata.

Reply:
> It would be nice if there were a way to avoid packing the files up three times, but I'm not sure there is one.

`git archive` and `meson dist` are trying to do fundamentally different things as a result of dist scripts existing. It will never be possible to reuse git's output directly (at least, impossible except as a micro-optimization for a subset of cases that always excludes the creation of python dist tarballs with sdist metadata.

It would be possible to uplift the meson-python specific changes into `meson dist`, of course. At the most basic level, it's "just" a kind of domain specific dist script.

Reply:
> Note that reproducibility is not necessarily a requirement of `meson dist`

It would be nice for `meson dist` to be reproducible if the dist scripts are reproducible, following a bit more closely what `git archive` does. Using the correct file metadata in the archive does not seem difficult to achieve, and it is at least a step in the right direction.

> Meson uses shutil.make_archive, which isn't well known for it's high degree of configurability

`shutil.make_archive()` has `owner` and `group` argument, that seem to fit at least part of the requirement.

For avoiding packing the files three times, I was more thinking about not having the lower layer pack them at all and have the upper layer do the packing in the way it likes, more than reusing the tarball generated by the lower layer. meson-python just needs to add a file to the archive, and this is possible without unpacking and repacking. The reason why meson-python nedds to unpack and repack the tarball is to fix the rood directory in the arvhive: the project name and version seen by Meson and meson-python are not necessarily the same.

Reply:
> that seem to fit at least part of the requirement.

"Part", eh. :D

Granted this is something that could be properly changed by giving up shutil.make_archive as a bad idea and rolling tarfile/zipfile manually.

> The reason why meson-python nedds to unpack and repack the tarball is to fix the rood directory in the arvhive: the project name and version seen by Meson and meson-python are not necessarily the same.

... although they probably should be, and for dynamic version it will be required.

Reply:
> although they probably should be, and for dynamic version it will be required.

I'm not sure they should. For project that only want to produce a Python packages, specifying the version only in `pyproject.toml` seems a perfectly valid thing to do, and for project that are more than a Python package, it seems reasonable to have the Python part have a different name than the whole. Introducing a restriction for them to be the same does not seem worth just to make it a bit easier to generate the sdist and it would be easier only if we fix `meson dist` to emit the right archive member metadata.

Reply:
I'm not saying it should be restricted, just that in a very large number of common cases they will be the same and won't need tweaking.

It is, at least, a useful micro optimization to check for (assuming no other issues).

As for versions, you can provide it in either or. But if you provide it in meson.build you can handle git versioning as well as using the version as a replacement string to insert into e.g. _version.py, so it is my (entirely personal) opinion that people should want to declare it as dynamic, because it is simply superior.